docs/contributing/999-rfcs/2024-02-15-resource-pinning.md

```diff
@@ -0,0 +1,246 @@
+- The pinfile is human-readable and editable
+- The pinfile contains the minimal information necessary to achieve the functional requirements
+- The pinfile is stable and predictable between compilations on different machines
+- CLI provides clear error messages and guidance for resolving deletion/moves of pinned resources
+
+### Cases
+
+For the following cases, assume a user has expressed their intent to protect a resource from deletion/moves somehow.
+The following cases should be handled by this RFC:
+
+- A pinned resource's id was changed in source (using `as`)
+- A pinned resource was moved to a different path in the resource tree (e.g. placed inside a different class)
+- An update to wing itself changed a pinned resource's path
+- An update to a dependency changed the id's within a resource it exports, and the user has pinned an instance of that resource
+
+## User Journey
+
+Sally Stateful creates a wing file with a bucket
+
+```wing
+// main.w
+bring cloud;
+
+new cloud.Bucket() as "CoolBucket";
+```
+
+They run `wing compile` and then `wing test`. Nice, looks like everything works.
+
+Running `wing compile -t tf-aws` also succeeds with the following output and a new file called main.w.pin.json appears:
+
+```shell
+# (output)
+[+pin] aws_s3_bucket.CoolBucket
+```
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "aws_s3_bucket.CoolBucket": {
+        "type": "@cdktf/provider-aws.s3bucket.S3Bucket",
+        "attributes": {
+          "bucket_prefix": "coolbucket-c87e9a4d-",
+          "force_destroy": false
+        }
+      }
+    }
+  }
+}
+```
+
+Sally deploys this to AWS and everything is great.
+
+Some time later, Sally decided to move the bucket into a new abstraction.
+
+```wing
+// main.w
+bring cloud;
+
+class MyBucket {
+  bucket: cloud.Bucket;
+
+  new() {
+    this.bucket = new cloud.Bucket();
+  }
+}
+
+let myBucket = new MyBucket();
+```
+
+`wing compile` and `wing test` still work as expected with simulator output.
+
+`wing compile -t tf-aws` however fails with an error message
+
+```shell
+# (output)
+[-pin] aws_s3_bucket.CoolBucket
+[+pin] aws_s3_bucket.MyBucket_AD8CE4AC
+
+Error: The following pinned resource has been moved or deleted:
+  - aws_s3_bucket.CoolBucket
+
+If moved, update the key in the pinfile to reflect the new path.
+If deleted, remove the entry from the pinfile.
+
+```
+> In the future we will add support for `wing pin mv` and `wing pin rm`
+
+This refactor changed the location of the bucket in the generated terraform, but more importantly the name/prefix of the bucket to be deployed.
+This change is a destructive operation that required replacement.
+
+For now, Sally wants to avoid this destruction so she updates the pinfile to keep attributes the same 
+but updates the key to reflect the new path in the generated terraform.
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "aws_s3_bucket.MyBucket_AD8CE4AC": {
```

Comment:
How does wing knows to associate this new name with the old name. In order to preserve the resource, the emitted Terraform must include the old name.

Reply:
I can add the old path as a field here. I had that previously in the RFC and you mentioned that wasn't needed so I thought you knew something that I didn't 😁